pkg/storegateway/indexheader/encoding/reader.go

```diff
@@ -139,9 +146,9 @@ func (f *FileReader) Len() int {
 	return f.length - f.pos
 }
 
-// close closes the underlying resources used by this FileReader. This method
-// is unexported to ensure that all resource management is handled by DecbufFactory
-// which pools resources.
-func (f *FileReader) close() error {
+// Close closes the underlying resources used by this FileReader.
+func (f *FileReader) Close() error {
+	bufferPool.Put(f.buf)
```

Comment:
I guess we need to reset the buffer state before returning it?

Reply:
We reset it when we retrieve it from the pool: `NewFileReader()` calls `Reset()`, which calls `ResetAt()`, which then resets the buffer: https://github.com/grafana/mimir/pull/3691/files#diff-2ae6b52cbe80bef6dd8daedc332df3984d2394f761480ff8aeb94cc3586888a8R66

Reply:
I would add a comment here to explain it.

Reply:
I've added a comment in 9e85f11.